File: bindings/java/src/main/java/com/strling/Compiler.java
package com.strling;

import com.strling.jvm.NativeClient;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/** Canonical compile-request conveniences over a supplied native client. */
public final class Compiler {
    private final NativeClient client;

    public Compiler(NativeClient client) {
        this.client = Objects.requireNonNull(client, "client");
    }

    public Object compile(Map<String, ?> request, Map<String, ?> targetProfile) {
        return client.compile(request, targetProfile);
    }

    public Object compile(Map<String, ?> request) {
        return client.compile(request);
    }

    public Object check(Map<String, ?> request, Map<String, ?> targetProfile) {
        return compile(request, targetProfile);
    }

    public Object check(Map<String, ?> request) {
        return compile(request);
    }

    /** Compatibility name returning canonical compile data, never a local AST. */
    public Object parse(String source, SourceCompileOptions options) {
        SourceCompileOptions selected = options == null
                ? SourceCompileOptions.builder().build() : options;
        return client.compile(sourceCompileRequest(source, selected,
                Collections.singletonList("semantic")), selected.getTargetProfile());
    }

    public Object parse(String source) {
        return parse(source, null);
    }

    /** Compatibility name requesting a canonical TargetArtifact. */
    public Object parseToArtifact(String source, SourceCompileOptions options) {
        Objects.requireNonNull(options, "options");
        if (options.getTargetProfile() == null
                || options.getTargetProfileReference() == null) {
            throw new IllegalArgumentException(
                    "parseToArtifact requires an exact target profile and reference");
        }
        return client.compile(sourceCompileRequest(source, options,
                Arrays.asList("semantic", "portability", "target_artifact")),
                options.getTargetProfile());
    }

    public Object describe() {
        return client.describe();
    }

    public Object inspectTargetProfile(Map<String, ?> targetProfile) {
        return client.inspectTargetProfile(targetProfile);
    }

    public static Map<String, Object> sourceCompileRequest(
            String source, SourceCompileOptions options, List<String> fallbackOutputs) {
        Objects.requireNonNull(source, "source");
        SourceCompileOptions selected = options == null
                ? SourceCompileOptions.builder().build() : options;
        String specificationVersion = selected.getSpecificationVersion();
        String frontendId = selected.getFrontendId();

        Map<String, Object> frontend = new LinkedHashMap<>();
        frontend.put("id", frontendId);
        frontend.put("dialect_version", selected.getFrontendVersion());

        Map<String, Object> content = new LinkedHashMap<>();
        content.put("kind", "inline");
        content.put("encoding", "utf-8");
        content.put("media_type", selected.getMediaType());
        content.put("text", source);

        Map<String, Object> document = new LinkedHashMap<>();
        document.put("contract_version", "1.0.0");
        document.put("source_id", selected.getSourceId());
        document.put("specification_version", specificationVersion);
        document.put("frontend", frontend);
        document.put("content", content);
        document.put("provenance", Collections.singletonMap("kind", "authored"));

        Map<String, Object> input = new LinkedHashMap<>();
        input.put("kind", "source");
        input.put("document", document);

        Map<String, Object> request = new LinkedHashMap<>();
        request.put("contract_version", "1.0.0");
        request.put("specification_version", specificationVersion);
        request.put("input", input);
        if (selected.getTargetProfileReference() != null) {
            request.put("target_profile", selected.getTargetProfileReference());
        }
        List<String> requested = selected.getRequestedOutputs() == null
                ? fallbackOutputs : selected.getRequestedOutputs();
        request.put("requested_outputs",
                Collections.unmodifiableList(new ArrayList<>(requested)));
        request.put("compiler_options", selected.getCompilerOptions());
        return Collections.unmodifiableMap(request);
    }
}

File: bindings/csharp/src/STRling/Canonical/Compiler.cs
using System.Text.Json;
using Strling.Native;

namespace Strling;

/// <summary>Canonical compile conveniences over a supplied native client.</summary>
public sealed class Compiler
{
    private readonly NativeClient client;

    public Compiler(NativeClient client) => this.client = client ?? throw new ArgumentNullException(nameof(client));

    public JsonElement Compile(JsonElement request, JsonElement? targetProfile = null) =>
        client.Compile(request, targetProfile);

    public JsonElement Check(JsonElement request, JsonElement? targetProfile = null) =>
        Compile(request, targetProfile);

    /// <summary>Compatibility name returning canonical compile data, never a local AST.</summary>
    public JsonElement Parse(string source, SourceCompileOptions? options = null)
    {
        var selected = options ?? new SourceCompileOptions();
        return Compile(SourceCompileRequest(source, selected, ["semantic"]), selected.TargetProfile);
    }

    /// <summary>Compatibility name requesting a canonical TargetArtifact.</summary>
    public JsonElement ParseToArtifact(string source, SourceCompileOptions options)
    {
        ArgumentNullException.ThrowIfNull(options);
        if (!options.TargetProfile.HasValue || !options.TargetProfileReference.HasValue)
            throw new ArgumentException("ParseToArtifact requires an exact target profile and reference", nameof(options));
        return Compile(
            SourceCompileRequest(source, options, ["semantic", "portability", "target_artifact"]),
            options.TargetProfile);
    }

    public JsonElement Describe() => client.Describe();
    public JsonElement InspectTargetProfile(JsonElement targetProfile) => client.InspectTargetProfile(targetProfile);

    public static JsonElement SourceCompileRequest(
        string source,
        SourceCompileOptions? options = null,
        IReadOnlyList<string>? fallbackOutputs = null)
    {
        ArgumentNullException.ThrowIfNull(source);
        var selected = options ?? new SourceCompileOptions();
        var mediaType = selected.MediaType;
        if (selected.FrontendId == "legacy_regex" && mediaType == "text/strling") mediaType = "text/x-regex";
        var request = new Dictionary<string, object?>
        {
            ["contract_version"] = "1.0.0",
            ["specification_version"] = selected.SpecificationVersion,
            ["input"] = new Dictionary<string, object?>
            {
                ["kind"] = "source",
                ["document"] = new Dictionary<string, object?>
                {
                    ["contract_version"] = "1.0.0",
                    ["source_id"] = selected.SourceId,
                    ["specification_version"] = selected.SpecificationVersion,
                    ["frontend"] = new Dictionary<string, object?>
                    {
                        ["id"] = selected.FrontendId,
                        ["dialect_version"] = selected.FrontendVersion,
                    },
                    ["content"] = new Dictionary<string, object?>
                    {
                        ["kind"] = "inline",
                        ["encoding"] = "utf-8",
                        ["media_type"] = mediaType,
                        ["text"] = source,
                    },
                    ["provenance"] = new Dictionary<string, object?> { ["kind"] = "authored" },
                },
            },
            ["requested_outputs"] = selected.RequestedOutputs ?? fallbackOutputs ?? ["semantic"],
            ["compiler_options"] = selected.CompilerOptions,
        };
        if (selected.TargetProfileReference.HasValue)
            request["target_profile"] = selected.TargetProfileReference.Value;
        return JsonSerializer.SerializeToElement(request);
    }
}
